OOModel/src/CodeGenerationVisitor.cpp

```diff
@@ -0,0 +1,178 @@
+#include "expressions/BooleanLiteral.h"
+
+namespace OOModel {
+
+CodeGenerationVisitor::CodeGenerationVisitor(QMap<QString, Model::Node *> args) : args_{args} {}
+
+void CodeGenerationVisitor::init()
+{
+	addType<ReferenceExpression>(visitReferenceExpression);
+	addType<Model::NameText>(visitNameText);
+	addType<MetaCallExpression>(visitMetaCallExpression);
+}
+
+void CodeGenerationVisitor::visitChildren(Model::Node* n)
+{
+	for (auto child : n->children()) visit(child);
+}
+
+void CodeGenerationVisitor::visitReferenceExpression(CodeGenerationVisitor* v, OOModel::ReferenceExpression* n)
+{
+	auto input = n->name();
+
+	if (!input.contains("##"))
+	{
+		if (auto argument = v->args_[input])
+		{
+			if (auto argumentReference = DCast<ReferenceExpression>(argument))
+			{
+				// case: there exists a replacement and it is another ReferenceExpression
+				// -> copy name (replacing it would cause child nodes of n to disappear)
+
+				n->setName(argumentReference->name());
+			}
+			else
+			{
+				// case: there exists a replacement and it is not a ReferenceExpression
+				// -> replace node
+
+				auto p = n->parent();
+
+				p->beginModification("change node in code generation visitor");
+
+				auto cloned = argument->clone();
+				p->replaceChild(n, cloned);
+
+				p->endModification();
+
+				// visit the cloned tree and return to avoid visiting the children of n
+				v->visitChildren(cloned);
+				return;
+			}
+		}
+	}
+	else
+	{
+		// case: n's name is a concatenated identifier (a##b)
+		// -> build identifier
+
+		QStringList parts = input.split("##");
+
+		for (auto i = 0; i < parts.size(); i++)
+			if (auto argument = DCast<ReferenceExpression>(v->args_[parts[i]]))
+				parts[i] = argument->name();
+
+		n->setName(parts.join(""));
+	}
+
+	v->visitChildren(n);
+}
+
+void CodeGenerationVisitor::visitNameText(CodeGenerationVisitor* v, Model::NameText* n)
+{
+	auto input = n->get();
+
+	if (!input.contains("##"))
+	{
+		if (auto argument = DCast<ReferenceExpression>(v->args_[input]))
+		{
+			// case: n's name is an existing argument of type ReferenceExpression
+			// -> copy name as new text
+
+			n->set(argument->name());
+		}
+	}
+	else
+	{
+		// case: n's text is a concatenated text (a##b)
+		// -> build text
+
+		QStringList parts = input.split("##");
+
+		for (auto i = 0; i < parts.size(); i++)
+			if (auto argument = DCast<ReferenceExpression>(v->args_[parts[i]]))
+				parts[i] = argument->name();
+
+		n->set(parts.join(""));
+	}
+
+	v->visitChildren(n);
+}
+
+void CodeGenerationVisitor::visitMetaCallExpression(CodeGenerationVisitor* v, MetaCallExpression* n)
+{
+	/*
+	 * process arguments before generating.
+	 * this ensures proper argument propagation.
+	 */
+	for (auto child : n->children()) v->visit(child);
+
+	if (!n->metaDefinition())
+	{
+		// case: unresolved meta definition
+		// -> check if it is a predefined meta function
+
+		if (auto metaDef = DCast<ReferenceExpression>(n->callee()))
```

Comment:
In that case nothing happens (assuming the misspelled name does not match a predefined function) since the handlePredefinedFunction procedure will just fall through all cases.